src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java

```diff
@@ -243,16 +221,19 @@ private int toDecimal(double v, FormattedFPDecimal fd) {
                 if (0 < mq & mq < P) {
                     long f = c >> mq;
                     if (f << mq == c) {
-                        return toChars(f, 0, fd);
+                        return toChars(str, index, f, 0, fd) - start;
                     }
                 }
-                return toDecimal(-mq, c, 0, fd);
+                return toDecimal(str, index, -mq, c, 0, fd) - start;
             }
             if (t != 0) {
                 /* subnormal value */
-                return t < C_TINY
-                       ? toDecimal(Q_MIN, 10 * t, -1, fd)
-                       : toDecimal(Q_MIN, t, 0, fd);
+                int dk = 0;
+                if (t < C_TINY) {
+                    t *= 10;
+                    dk = -1;
+                }
+                return toDecimal(str, index, Q_MIN, t, dk, fd) - start;
```

Comment:
I suggest restoring the original logic like so:
```
                /* subnormal value */
                return (t < C_TINY
                        ? toDecimal(str, index, Q_MIN, 10 * t, -1, fd)
                        : toDecimal(str, index, Q_MIN, t, 0, fd)) - start;
```

Reply:
I like the new implementation, the code is cleaner, is your suggestion to revert to the original version due to smaller changes?

Reply:
In principle, you should not arbitrarily change code that is correct, and only limit your changes to reach the goal of the PR.
My suggestion is the minimal change required, yours is more than strictly needed.

Reply:
Got it, should I change it back?

Reply:
If there's nothing else, no, there's no need to switch back.

Sorry if I sound pedantic on some principles that I feel are important for the good health of OpenJDK ;-)